lib/scrapers/gmailscrape.py

```diff
@@ -0,0 +1,189 @@
+from tornado import gen
+from tornado.httpclient import HTTPError
+
+import httplib2
+import base64
+from email.parser import BytesParser as EmailParser
+from email import policy
+from lib.config import CONFIG
+import random
+
+from apiclient.discovery import build
+from oauth2client import client
+
+
+def equalize_dict(data, max_num):
+    """
+    Quick and dirty code to randomly sample from a dictionary of lists in order
+    to roughly equalize their lengths so that IN TOTAL we have `max_num`
+    elements.
+    NOTE: this will randomly shuffle the values of the data so don't expect it
+    to maintain any oder
+    """
+    max_per_bin = max_num // len(data)
+    to_kill = sum(map(len, data.values())) - max_num
+    while to_kill > 0:
+        for key, values in data.items():
+            if len(values) <= max_per_bin:
+                continue
+            dv = min(to_kill, len(values) - max_per_bin)
+            killing = random.randint(1, int(dv))
+            random.shuffle(values)
+            for i in range(killing):
+                values.pop()
+            to_kill -= killing
+            if to_kill <= 0:
+                break
+    return data
+
+
+class GMailScraper(object):
+    name = 'gtext'
+
+    def __init__(self):
+        self.tokens = []
+        with open('lib/scrapers/snippets.txt', 'r') as fd:
+            self.tokens = [s.strip() for s in fd]
+
+    @property
+    def num_threads(self):
+        if CONFIG['_mode'] == 'prod':
+            return 1000
+        return 100
+
+    def get_content(self, raw):
+        data = base64.urlsafe_b64decode(raw)
+        email_parser = EmailParser(policy=policy.default)
+        email = email_parser.parsebytes(data)
+        plain = email.get_body(preferencelist=('plain',))
+        body = None
+        if plain:
+            body = plain.get_payload()
+        email_dict = dict(email)
+        email_dict['body'] = body
+        return email_dict
+
+    def paginate_messages(self, service, response, max_results=None):
+        threads = []
+        if 'messages' in response:
+            for i in response['messages']:
+                threads.append(i['threadId'])
+
+        while 'nextPageToken' in response:
+            page_token = response['nextPageToken']
+            response = service.users().messages().list(
+                userId='me',
+                pageToken=page_token).execute()
+            for i in response['messages']:
+                threads.append(i['threadId'])
+            if max_results and len(threads) >= max_results:
+                break
+        return threads
+
+    def get_recipient(self, email_data):
```

Comment:
Something's wrong here and it's only getting the first letter of the recipient (example from response: ` 'people': ['m']`)